### PR TITLE
Version policy

### PR DESCRIPTION
**Pull request by paulidale**
Add a versioning policy based on the web pages.

Also begin a definitions policy document.

This is a follow on from #4 which was branched in the wrong place.
All feedback there has been addressed.

### CODE REVIEW

File: policies/definitions.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+## Public interface
+
+A public interface is any function, global variable, structure or macro
+declared in a public header file.
+Refer to the [API compatibility policy] for specific details.
+
+## Stable release
+
+A stable release is a patch release from an existing supported minor
+release branch.
```
Comment (levitte) [inline policies/definitions.md:84]: Not sure I understand how this differs from "Patch release", defined above.

Also, I think that we have commonly called this a "Release update"...

Comment (paulidale) [reply policies/definitions.md:84]: We use the term _stable release_ in our policies, it needs to be defined.  Here is where it is defined.

The document, perhaps, should also define _release update_.  We don't appear to have used that term (yet) in the policy repos.

Comment (levitte) [reply policies/definitions.md:84]: Yeah, and actually, this definition isn't quite it.  A stable release *isa* minor or major release (including all updates, present and future), i.e. 3.0.0 is a stable release (or one could say 3.0.x).

Comment (paulidale) [reply policies/definitions.md:84]: This is the definition from our [stable release policy](https://github.openssl.org/otc/technical-policies/blob/master/policies/stable-release-updates.md#definitions), which I'll note you voted in favour of :grin:

3.0.0 is technically a patch release (as well as a minor and a major one).  From this perspective, the statement is correct.

Comment (levitte) [reply policies/definitions.md:133]: Damn it, I didn't notice it then. I find this description quite confusing, and your attempt at retrofitting an explanation doesn't help.

Comment (paulidale) [reply policies/definitions.md:133]: Suggest what the glossary should say about it.

We can and should more rigorously define the term in the stable release policy.

Comment (paulidale) [reply policies/definitions.md]: Fixed it here.

Comment (levitte) [reply policies/definitions.md]: I've submitted an amendment PR for the stable release policy.  openssl/technical-policies#24

File: policies/versioning-policy.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+Both the API and ABI will remain compatible across patch releases.
+
+> For example, a program linked with OpenSSL release 3.0.0 can
+> run with OpenSSL 3.0.1 without changes.
+
+Exceptions to these rules require a vote by the OMC.
+
+## Long term stable release
+
+A release can be designated as a release.  LTS releases will be supported
```
Comment (mattcaswell) [inline policies/versioning-policy.md:71]: "A release can be designated as a release." I think you are missing LTS from that sentence.

File: policies/versioning-policy.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+for at least five years and the project will specify a LTS release at
```
Comment (mattcaswell) [inline policies/versioning-policy.md:72]: an LTS

File: policies/versioning-policy.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+least every four years.  Non-LTS releases will be supported for at least
+two years.
```
Comment (mattcaswell) [inline policies/versioning-policy.md:74]: The title of this section is "Long term stable releases". But it actually talks about the support period for non-LTS releases too. Also the para below is applicable to both LTS and non-LTS releases.

File: policies/versioning-policy.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+During the final year of support, we do not commit to anything other
+than security fixes.  Before that, bug and security fixes will be applied
+as appropriate.
+
+The addition of new platforms to LTS branches is acceptable so long as
+the required changes consist solely of additions to configuration.
```
Comment (mattcaswell) [inline policies/versioning-policy.md:81]: Does this also apply to non-LTS branches?

File: policies/versioning-policy.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+- Letter releases, such as 1.0.2a, exclusively contain bug and security
+  fixes and no new features.
+- Releases that change the last digit, e.g. 1.1.0 vs. 1.1.1, can and
+  are likely to contain new features, but in a way that does not break
+  binary compatibility.  This means that an application compiled and
+  dynamically linked with 1.1.0 does not need to be recompiled when the
+  shared library is updated to 1.1.1.
+- Releases that change the second digit, e.g. 1.0.0 vs 1.1.0, break
+  both ABI and API compatibility.  The primary instance of this was
+  the transition to opaque internal structures occurred with OpenSSL
```
Comment (mattcaswell) [inline policies/versioning-policy.md:117]: that occurred with?

File: policies/versioning-policy.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  release 1.1.0.
+
+It should be noted that some features are transparent to the application
+such as the maximum negotiated TLS version and cipher suites, performance
+improvements and so on.  There is no need to recompile applications to
+benefit from these features.
```
Comment (mattcaswell) [inline policies/versioning-policy.md:123]: I'm not really sure what the intent of this para is.

File: policies/definitions.md [hunk 8]
```diff
@@ -0,0 +1,150 @@
+# Glossary of OpenSSL terms and definitions
```
Comment (mattcaswell) [inline policies/definitions.md:1]: Isn't "definitions" kind of implicit in Glossary? i.e. is "Glossary of OpenSSL terms" enough? Or alternatively "Definition of OpenSSL terms"

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: policies/definitions.md (hunk 8 above)
Comment (paulidale) [reply policies/definitions.md:1]: I prefer glossary, this doesn't really define anything it, instead, provides a brief description and a reference to where the term is defined.

File: policies/versioning-policy.md (hunk 7 above)
Comment (paulidale) [reply policies/versioning-policy.md:123]: Me either.  It's taken entirely from the description of earlier versioning in the original web page.

I can remove?

File: policies/versioning-policy.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+> as efficient as it was.
+
+In order to maintain stability and limit rework across major versions:
+
+- Existing [public interfaces] will remain unmodified except where changes
+  are unlikely to break source compatibility or where structures are being
+  made opaque.
+- No existing public interface can be removed until its replacement has
+  been in place in an LTS stable release. The original interface must
+  also have been documented as deprecated for at least 5 years.
```
[the remark this answers is not in the document]
Comment (paulidale) [reply policies/versioning-policy.md:38]: I'm not proud, I've swapped your wording in.

File: policies/versioning-policy.md (hunk 4 above)
Comment (paulidale) [reply policies/versioning-policy.md:74]: I'll put a "supported releases" section in, although the flow will still be a bit odd.

File: policies/glossary.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
+# Glossary of OpenSSL terms
+
+## ABI
+
+Application binary interface.
+
+An ABI compatible release allows applications to
+be relinked against the new library with the expectation that everything will
```
Comment (kroeckx) [inline policies/glossary.md:8]: run? If it's a shared object, you can just update the library without relinking it.

File: policies/glossary.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+An [alpha release] is an early pre-release version.
+Refer to the [versioning policy] for specific details.
+
+## API
+
+Application programming interface.
+
+An API compatible release allows application to be recompiled against the new
+library with the expectation that everything will continue to function.
```
Comment (kroeckx) [inline policies/glossary.md:21]: Is that an API compatible but non-ABI compatible release?

File: policies/glossary.md [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+## CCLA
+
+<a name="ccla"></a>_Corporate Contributor Licence Agreement_
+
+The [CCLA] is a legal document that grants certain rights to code
+contributed as part of the employment of the contributor.  All non-trivial
+contributions on behalf of an employer must be submitted under a CCLA.
+
+qv [ICLA](#icla)
```
Comment (kroeckx) [inline policies/glossary.md:44]: Just say "Also see"?

File: policies/glossary.md [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+## Committer
+
+OpenSSL committers are contributors who have commit access to the OpenSSL
+source code repository.
+
+## End-user documentation
+
+End-user documentation is documentation intended for users of the OpenSSL
+libraries and commandline utilities.  Refer to the [stable release update
+policy] for specific details.
```
Comment (kroeckx) [inline policies/glossary.md:71]: I'm unsure why pointing to the stable release update here is relevant.

File: policies/versioning-policy.md (hunk 7 above)
Comment (mattcaswell) [reply policies/versioning-policy.md:128]: I'd say remove it.

Comment (paulidale) [reply policies/versioning-policy.md:128]: Removed.

Comment (mattcaswell) [reply policies/versioning-policy.md:128]: I think you pushed this fix to the wrong branch. This wording has gone from #8 but not from this PR.

Comment (paulidale) [reply policies/versioning-policy.md:128]: You're correct.  #8 is based over this one and I've muddled things a few times now.

**Comment by paulidale** [comment]
Calling for a vote on this policy.

**Comment by paulidale** [comment]
Vote: [+1]

File: policies/glossary.md [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+
+The [ICLA] is a legal document that grants certain rights to your
+contributed code.  All non-trivial contributions must be submitted under
+an ICLA.
+
+Also see: [CCLA](#ccla)
+
+## LTS
+
+[Long term support].  These releases are supported for longer than normal
```
Comment (tom-cosgrove-arm) [inline policies/glossary.md:87]: Should this be "_Long term support_" on a line on its own, like "_Individual Contributor Licence Agreement_" and "_OpenSSL Management Committee_"

File: policies/versioning-policy.md [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+> version of _0_ and a _patch_ version of _1_.
+
+This closely aligns with the expectations of users who are familiar
+with _semantic versioning_.  However, it is not quite _semantic versioning_
+because adopting _semantic versioning_ would mean changing our current
+Long Term Support ([LTS]) policies and practices.
+
+## Major release
+
+A major release is indicated by changing the _first digit_ of the version.
```
Comment (tom-cosgrove-arm) [inline policies/versioning-policy.md:23]: Using "first digit", "second digit" and "final digit" precludes versions such as `3.2.10` and `3.10.0`. Would "number" work better?

File: policies/glossary.md [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+Significant modifications for the OpenSSL library or commands are
+documented in the [CHANGES.md] file.
+
+## CLA
+
+_Contributor Licence Agreement_
+
+This is a legal document that grants certain rights to your contributed
+code.  All non-trivial contributions must be submitted under a CLA.
+Either an ICLA for personal contributions or both an ICLA and a CCLA
```
Comment (tom-cosgrove-arm) [inline policies/glossary.md:60]: (minor grammar) `Either...` isn't a full sentence, suggest replacing the period/full stop at the end of line 59 with a comma, leading to "All non-trivial contributions must be submitted under a CLA, either an ICLA for personal contributions or both an ICLA and a CCLA for work related contributions."

File: policies/glossary.md (hunk 14 above)
Comment (paulidale) [reply policies/glossary.md:87]: Yes, it should be.  I switched the formatting part way through and missed this one.

**Comment by paulidale** [comment]
Since I'm the only person who voted and since some issues have been raised, I'll fix them and restart the vote.

**Comment by paulidale** [comment]
Vote: Accept openssl/general-policies PR#6 - Version policy as of commit ee22052. This will become an official OMC policy.

**Comment by paulidale** [comment]
Vote: [+1]

File: policies/versioning-policy.md (hunk 15 above)
Comment (paulidale) [reply policies/versioning-policy.md]: I've changed to "number" throughout and got the italics consistent.  Also changed a "last digit" to "final number".

**Comment by t-j-h** [comment]
Vote: [+1]

**Comment by t-j-h** [comment]
The glossary definition of beta really doesn't capture the essential elements of the versioning definition of beta. I think that can lead to confusion as there are varying views on what beta means - is it just another round of alpha-like releases or is it effectively able to be a release candidate. We had mixed usage of "beta" in previous release handling (much closer to the former than the later for releases before 3.0 and much closer to the latter than the former for 3.0). The glossary should note that as well IMHO.

Note this is something that can be adjusted later on - hence the +1 vote for the text as-is.

**Comment by paulidale** [comment]
This glossary is deliberately _not_ defining anything.  It offers a brief description and a link to the formal definition.

A note to this effect at the beginning would be a worthwhile addition.

**Comment by kroeckx** [comment]
I'm voting +1

**Comment by paulidale** [comment]
The vote passes.

```
opic: Accept openssl/general-policies PR#6 - Version policy as of
       commit ac8266d. This will become an official OMC policy.
Proposed by: pauli
Issue link: https://github.com/openssl/general-policies/pull/6
Public: yes
Opened: 2022-02-09
Closed: 2022-02-17
Accepted:  yes  (for: 4, against: 0, abstained: 0, not voted: 2)

  Kurt       [+1]
  Mark       [  ]
  Matt       [+1]
  Pauli      [+1]
  Richard    [  ]
  Tim        [+1]
```

Merged this policy.

**Comment by iamamoose** [comment]
[+1]

**Pull request closed without merging**